### DISCUSSION TITLE
Transitioning from HDDM to HSSM: Specifying between and within subject variables in the HSSM

### DISCUSSION THREAD

**Discussion by Billygitzhu** [lnccbrown/HSSM#287] [Q&A]
Hello everyone, I'm new to the HSSM package, so please excuse my questions if they seem less specialized.

In the HSSM documentation: 
> As a previous HDDM user, you may recognize that now proper mixed-effect models are viable!
> You should be able to handle between and within subject effects naturally now!

**Experiment Background**
However, there remains a lingering question about specifying category variables in this context.
I'm currently working with a dataset that involves three independent variables. One of these variables is a between-subjects variable Type (A, B), while the other two are within-subjects variables with Magnitude (none, small, large), and Valence (reward, punishment).

**Question**
My main inquiry is how to specify the model in HSSM to account for these variables. 

In the past, during the HDDM era, we had options such as using hddm.HDDM() with depends_on or employing hddm.HDDMRegressor(). Given the mixed design of my study, it seems more appropriate to use hddm.HDDMRegressor. However, I have problem about this choice because the output does not seem to provide parameter values for each subject under different conditions.
For instance, using depends_on={'v':['type','magnitude','valence']} in hddm.HDDM(), we could obtain parameter values (v) for each condition and each subject within the combinations of magnitude and valence.
Conversely, when using hddm.HDDMRegressor(), the output only displays parameters for each subject across the entire experiment, without breakdowns for individual conditions.

The reason need to get the parameters for each subject at each level of the variables is that finding the interaction effect between two group,  so I need the parameters for each subject in the different conditions to use the ANOVA.

**How to solve problem in HSSM**
As for my previous question in HDDM, if I opt for hddm.HDDMRegressor, how should I structure my variables? 
For now, in HSSM, I wonder how to best specify the model in this framework. Additionally, does the specified model output the parameters for each subject at each level of the categorical variables?

I appreciate any insights and guidance on this matter. Thank you!

Billy

**Comment by AlexanderFengler** [comment]
HSSM allows you to use the wilcoxon formula notation to specify regression models, which allow for random effects across any categorical variable. 

You can check the [main tutorial in the documentation](https://lnccbrown.github.io/HSSM/tutorials/main_tutorial/#simulate-data) for some examples, but you might also find the [Bambi documentation](https://bambinos.github.io/bambi/notebooks/sleepstudy.html) useful for some broader examples on how to use the formula notation (HSSM builds on Bambi to process general linear models).

Finally, Bambi in turn builds on the [formulae](https://bambinos.github.io/formulae/) package specifically to build design-matrices from model strings.

In case you are familiar with the [lme4](https://cran.r-project.org/web/packages/lme4/index.html) package in R for mixed-effects regression, you are essentially able to specify model formulas the same way as in this widely used package. 

Specifically to your question, you may want to try something like this,

`v ~ 1 + (1|subj_idx) + (1|category)`

which should create category wise and subject wise random effects (so one parameter per subject for each category, or conversely for each subject one parameter per category).

Let us know if this doesn't help.

Alex

**Reply by Billygitzhu** [reply]
Alex,
Hello!

Thank you for your useful response, I ran my model at colab based on your suggestion.
However, after running it there are still a couple of questions that are not very clear at the moment.

**1. How to deal with between-subject variable?** 
Should I split the different groups of subjects into two model calculations? For example, group A does it once and then group B re-runs the model once. If not, how do I join the group variable in the model?

**2. Did formula setup correct?** 
Based on your suggestion, I have set up my model as follows
```
model1 = hssm.HSSM(data=df, model='ddm',
               include=[
                  {
                    "name":"v",
                    "formula":"v ~ 1 + (1|subject) + (1|stimulus)",
                  }
               ])
infer_reg_model1 = model1.sample()
az.summary(infer_reg_model1).to_excel('result1.xlsx')
```



If it is correct, how should I interpret the result1?
_The following graph shows the results_
![image](https://github.com/lnccbrown/HSSM/assets/111554847/637e93d7-a0f4-45af-b338-feb0cf664c8b)

If I need to know the 'v' of each SUBJECT in every single condition, what should I do?
For example, I need to know the 'v' of subject1 at 'loss1','loss5','loss10'... , how should I get it?
**My primary goal is that I would like to get the 'v' for each SUBJECT in each condition**, so that I can use ANOVA to calculate whether there is a significant difference and interaction effect between groups.


So, to accomplish the idea, I modified the code
```
model2 = hssm.HSSM(data=df, model='ddm',
               include=[
                  {
                    "name":"v",
                    "formula":"v ~ 1 + (stimulus|subject)",
                  }
               ])
infer_reg_model2 = ddm_model_model2.sample()
az.summary(infer_reg_model2).to_excel('result2.xlsx')

```
_The following graph shows the results_
![image](https://github.com/lnccbrown/HSSM/assets/111554847/11181c47-c304-46e3-9a80-3be8d70b3d22)


**It seemed to contain a 'v' for each subject in each condition.**
However, I am not sure if this is correct and used it directly for ANOVA.



Last question is 
**If I then add a new within-subjects variable, I also need to find the interaction effects between the variable and exist variables. How should I add this variable to the new model?**

Thanks again!

Best wishes,
Billy

**Comment by moxtoby** [comment]
Hello Billy,

I am also just transitioning from HDDM to HSSM, and finding it difficult to work out the syntax to define between subjects and within subjects variables. I was hoping you have some breakthrough and can share how you defined your variables? I have 1 between subjects variable (group:3 levels), and one within subjects variable (condition:2 levels). I can't seem to see an example of a mixed design in the HSSM tutorial to follow. 
By the way, I am not sure if the output of the DDM parameters are suitable for ANOVA, as each variable is not "independent" as they have been calculated based on group performance. 
Thanks in advance.

**Reply by Billygitzhu** [reply]
I apologize for not following your response in time recently.

I have recently used HDDM for my assignment and as far as I am concerned in the literature on the subject, there exists a test of difference between groups using the HDDM package to fit the modeled data after obtaining the metrics (eg. v) after taking ANOVA.

I would think that you could use depends_on directly in the HDDM package for your within-group differences and split the between-group data into two groups (depending on the number of groups, e.g., three) for two fits (depending on the number of groups you have).


`
hddm.HDDM(group1_data, include={'a','v','z','t'}, dependes_on={'a':{conditions},'v':{conditions})`
`
hddm.HDDM(group2_data, include={'a','v','z','t'}, dependes_on={'a':{conditions},'v':{conditions})`

If you have any questions feel free to continue the discussion.

Best wishes,
Billy

**Reply by moxtoby** [reply]
Thank you very much Billy.
After a bit of trial and error, I have also come to the same conclusion, to run separate HDDM commands for each group. Thank you again for your help.

Best wishes,
Michelle
